src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+    overscan: 75,
+    scrollMargin: codeDisplayOverlayRef.current?.offsetTop ?? 0,
+  })
+
+  const div = codeDisplayOverlayRef.current
+  if (div) {
+    // set the parent div height to the total size of the virtualizer
+    div.style.height = `${virtualizer.getTotalSize()}px`
+    div.style.position = 'relative'
+  }
+
+  // TODO - finalize this logic
+  // scroll to line on the initial render
+  useEffect(() => {
+    if (initialRender.current) {
+      initialRender.current = false
+      const index = parseInt(location.hash.slice(2), 10)
+      virtualizer.scrollToIndex(index)
+    }
+  }, [location.hash, virtualizer])
+
+  return (
+    <div className="flex flex-1" ref={wrapperRef}>
+      {/* this div contains the line numbers */}
+      <div className="relative z-[2] h-full w-[82px] min-w-[82px] pr-[10px]">
+        {virtualizer.getVirtualItems().map((item) => (
+          <div
+            ref={virtualizer.measureElement}
+            key={item.index}
+            data-index={item.index}
+            style={{
+              height: `${item.size}px`,
+              transform: `translateY(${
+                item.start - virtualizer.options.scrollMargin
+              }px)`,
+            }}
+            className={cn(
+              `absolute left-0 top-0 w-full border-r border-ds-gray-tertiary bg-white px-4 text-right text-ds-gray-quaternary hover:cursor-pointer hover:text-black`,
+              location.hash === `#L${item.index}` &&
+                'font-semibold text-ds-gray-quinary',
+              coverage[item.index] === 'H' && 'bg-ds-coverage-covered',
+              coverage[item.index] === 'M' &&
+                'bg-ds-coverage-uncovered after:absolute after:inset-y-0 after:right-0 after:border-r-2 after:border-ds-primary-red',
+              coverage[item.index] === 'P' &&
+                'bg-ds-coverage-partial after:absolute after:inset-y-0 after:right-0 after:border-r-2 after:border-dotted after:border-ds-primary-yellow'
+            )}
+            onClick={() => {
+              if (location.hash === `#L${item.index}`) {
+                location.hash = ''
+              } else {
+                location.hash = `#L${item.index}`
+              }
+              history.push(location)
+            }}
+          >
+            {location.hash === `#L${item.index}` ? '#' : null}
+            {item.index}
+          </div>
+        ))}
+      </div>
+      {/* this div contains the actual code lines */}
+      <div
+        // @ts-expect-error - TODO - Update inert when inert is available in React 19
+        inert=""
+        className="pointer-events-none size-full"
+      >
+        {virtualizer.getVirtualItems().map((item) => {
+          // get any specific things from code highlighting library for this given line
+          const { style: lineStyle } = getLineProps({
+            // casting this cause it is guaranteed to be present in the array
+            line: tokens[item.index] as Token[],
```

Comment:
Would this do the same thing?
```suggestion
            line: tokens[item.index]!,
```